include/image-commands.mk

```diff
@@ -217,6 +217,96 @@ define Build/append-kernel
+# https://stackoverflow.com/a/5887751
+define newline
+
+
+endef
+
+define Build/append-raw-dtb-to-elf-linker-script
+OUTPUT_ARCH(ARCH_LD)
+SECTIONS {
+	/*
+	 * Preprocess the load address
+	 * -Ttext $ADDR is brittle:
+	 * https://lkml.kernel.org/lkml/20200413153453.zi4jvu3c4ul23e23@google.com/
+	 */
+	. = ZLOADADDR;
+	.text : {
+		startup = .;
+		*(.text)
+		*(.text.*)
+		*(.data)
+		*(.data.*)
+	}
+}
+endef
+
+# Replaces kernel-bin
+# requires ELF kernel
+# uses $@: target_rule_name example: /**/build_dir/target-*/linux-*/tmp/openwrt-*{,-initramfs}-kernel.bin
+# uses $<: first_prerequisite_name example: /**/build_dir/target-*/linux-*/$KERNEL_NAME{,-initramfs}
+# process:
+# copy the $KERNEL_NAME to $KDIR/tmp
+# convert to bytecode
+# append DTB
+# preprocess linker script https://elixir.bootlin.com/linux/v5.4.66/source/scripts/Makefile.build#L353
+# convert back to ELF (which requires data from the original ELF)
+# - ELF type example: elf32-tradlittlemips
+# - load address example: 0x80931000
+define Build/append-raw-dtb-to-elf
+	$(eval $@ : KERNEL_ELF_INPUT := $<)
+	$(eval $@ : KERNEL_OUTPUT := $@)
+	$(eval $@ : KERNEL_TEMP_UNLINKED := $(KERNEL_OUTPUT)_unlinked)
+	$(eval $@ : KERNEL_TEMP_BC := $(KERNEL_OUTPUT)_binary)
+	$(eval $@ : LINKER_SCRIPT := $(KERNEL_OUTPUT).lds)
+	$(eval $@ : LINKER_SCRIPT_UNPROCESSED := $(LINKER_SCRIPT).S)
+
+	$(eval $@ : ZLOADADDR := $(filter-out Entry point address:, \
+				   $(shell readelf -h $(KERNEL_ELF_INPUT) | \
+				     grep Entry)))
+	$(eval $@ : ELF_FORMAT := $(filter-out $(KERNEL_ELF_INPUT): file format, \
+				    $(shell $(TARGET_CROSS)objdump -h $(KERNEL_ELF_INPUT) | \
+				      grep 'file format' )))
+
+	$(eval $@ : ARCH_LD := $(subst mipsel,mips,$(ARCH)))
+
+	$(eval $@ : TEMP_FILENAMES := $(KERNEL_TEMP_UNLINKED) \
+				      $(KERNEL_TEMP_BC) \
+				      $(LINKER_SCRIPT) \
+				      $(LINKER_SCRIPT_UNPROCESSED))
+
+	$(info KERNEL_OUTPUT=$(KERNEL_OUTPUT))
+	$(info KERNEL_ELF_INPUT=$(KERNEL_ELF_INPUT))
+	$(info ZLOADADDR=$(ZLOADADDR))
+	$(info ELF_FORMAT=$(ELF_FORMAT))
+	$(info ARCH_LD=$(ARCH_LD))
+
+	rm -f $(KERNEL_OUTPUT)
+	rm -f $(TEMP_FILENAMES)
+
+	$(TARGET_CROSS)objcopy -O binary \
+			       -R .reginfo -R .note -R .comment -R .mdebug -R .MIPS.abiflags -S \
+			       $(KERNEL_ELF_INPUT) $(KERNEL_TEMP_BC)
+	cat $(KDIR)/image-$(firstword $(DEVICE_DTS)).dtb >> $(KERNEL_TEMP_BC)
+
+	$(TARGET_CROSS)ld -r -b binary --oformat="$(ELF_FORMAT)" \
+			  -o $(KERNEL_TEMP_UNLINKED) $(KERNEL_TEMP_BC)
+
+	printf '$(subst $(newline),\n,$(Build/append-raw-dtb-to-elf-linker-script))' \
+		  > $(LINKER_SCRIPT_UNPROCESSED)
+	$(TARGET_CROSS)cpp -P -U$(ARCH_LD) -U$(ARCH) \
+			   -DZLOADADDR="$(ZLOADADDR)" -DARCH_LD="$(ARCH_LD)" \
+			   -D__ASSEMBLY__ -DLINKER_SCRIPT -o $(LINKER_SCRIPT) \
+			   $(LINKER_SCRIPT_UNPROCESSED)
+	$(TARGET_CROSS)ld -e startup -T $(LINKER_SCRIPT) -z max-page-size=4096 \
+			  -o $(KERNEL_OUTPUT) $(KERNEL_TEMP_UNLINKED)
+
+	rm -f $(TEMP_FILENAMES)
+endef
```

Comment:
That's quite a long image recipe due to the problem I mentioned above:

> The remaining problem:
> kernel calculates the load address for zboot according to kernel size.
> How could we get this address and use it in image Makefile?

OpenWrt tries to break image building to smaller steps. But that's not achievable if we need to extract load address from elf (the first step) and use it in image packaging (the last step).

I'm considering adding the relocation code from our lzma-loader so that the zboot image can be loaded anywhere.
I'll post updates here if I managed to achieve that.

Reply:
> ... the relocation code from our lzma-loader ...

FYI, that awkward DIR-860L needs `relocate-kernel` recipe too if the image is packed with `seama` (even if it's LZMA Loader).
But that relocate recipe isn't needed if the image isn't packed with `seama`: the device happily boots plain `uImage` packed images.

Reply:
> I'm considering adding the relocation code from our lzma-loader so that the zboot image can be loaded anywhere.

Please don't. That code is not portable, meaning there will be a need to implement it for each architecture. That code also duplicates the kernel's own relocation code. As soon as you add that code (which is by the way what the "relocate" tool in target/linux/generic/image/relocate implements), you are already going down the rabbit hole of implementing a new loader all over again.

Please let's not stray from the goal of booting vanilla kernel.

Reply:
The single define here can be broken into as many parts as you like, by using the makefile variables that are already set in this define, then piping the defines together, as is usually done in image/$target.mk `KERNEL :=`?

Just let me know what you think is appropriate?
1. get-elf-information
2. output-bytecode
3. append-dtb (could use existing append-dtb recipe if happy to do the operation on the kernel output filename `$@`?)
4. output-elf
5. preprocess-ldscript
6. arrange-elf

I don't think we would even need to remove the local-scope-variable indicator `$@:`, as piped (split) definitions would be called by the one recipe `$@`?


Example, split define for retrieving the ELF information:
```
define Build/append-raw-dtb-to-elf-get-elf-information
	$(eval $@ : KERNEL_ELF_INPUT := $<)
	$(eval $@ : ZLOADADDR := $(filter-out Entry point address:, \
				   $(shell readelf -h $(KERNEL_ELF_INPUT) | \
				     grep Entry)))
	$(eval $@ : ELF_FORMAT := $(filter-out $(KERNEL_ELF_INPUT): file format, \
				    $(shell $(TARGET_CROSS)objdump -h $(KERNEL_ELF_INPUT) | \
				      grep 'file format' )))
	$(eval $@ : ARCH_LD := $(subst mipsel,mips,$(ARCH)))
endef
```
This just sets these recipe-scope makefile variables, which can be accessed by other _define_ piped later in the same recipe?

I do not want to add non-kernel code to the boot process if it can be avoided.
I started this because of the boot-issue I was seeing on mt7621 mikrotik with lzma-loader.


Cheers

Reply:
As I'm not familiar with openwrt buildroot, that's the easiest way I could think of to get rid of this long image recipe.
It would be better if someone could propose other ideas to deal with this problem.

Reply:
@john-tho 

That's great!
I think it could contain 4 steps:
1. extract load address from elf
2. copy binary code out
3. use the existing append-dtb
4. package bin into elf